vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml

```diff
@@ -125,91 +152,141 @@
-            sed -i 's/description: KubeadmConfig/description: upgradedKubeadmConfig/' \
-                {{working_dir}}/dev-repository//bootstrap-kubeadm/{{CAPI_REL_TO_VERSION}}/bootstrap-components.yaml
-            sed -i 's/description: KubeadmControlPlane/description: upgradedKubeadmControlPlane/' \
-                {{working_dir}}/dev-repository//control-plane-kubeadm/{{CAPI_REL_TO_VERSION}}/control-plane-components.yaml
-            sed -i 's/\bm3c\b/m3c2020/g' \
-                {{working_dir}}/dev-repository/infrastructure-metal3/{{CAPM3_REL_TO_VERSION}}/infrastructure-components.yaml
-            sed -i 's/\bm3c\b/m3c2020/g' \
-                {{working_dir}}/overrides/infrastructure-metal3/{{CAPM3_REL_TO_VERSION}}/infrastructure-components.yaml
-   
+    ansible.builtin.replace:
+      path: "{{ item.path }}"
+      regexp: "{{ item.regexp }}"
+      replace: "{{ item.replace }}"
+    loop:
+      - path: "{{working_dir}}/dev-repository/cluster-api/{{CAPI_REL_TO_VERSION}}/core-components.yaml"
+        regexp: 'description: Machine'
+        replace: "description: upgradedMachine"
+      - path: "{{working_dir}}/dev-repository//bootstrap-kubeadm/{{CAPI_REL_TO_VERSION}}/bootstrap-components.yaml"
+        regexp: 'description: KubeadmConfig'
+        replace: "description: upgradedKubeadmConfig"
+      - path: "{{working_dir}}/dev-repository//control-plane-kubeadm/{{CAPI_REL_TO_VERSION}}/control-plane-components.yaml"
+        regexp: 'description: KubeadmControlPlane'
+        replace: "description: upgradedKubeadmControlPlane"
+      # TODO: Should we use something more generic than "m3c2020"? Say "m3cnext" or "upgm3c"?
+      - path: "{{working_dir}}/dev-repository/infrastructure-metal3/{{CAPM3_REL_TO_VERSION}}/infrastructure-components.yaml"
+        regexp: '\bm3c\b'
+        replace: "m3c2020"
+      - path: "{{working_dir}}/overrides/infrastructure-metal3/{{CAPM3_REL_TO_VERSION}}/infrastructure-components.yaml"
+        regexp: '\bm3c\b'
+        replace: "m3c2020"
+
   - name: Perform upgrade on the target cluster
-    shell: |
-           # Add missing environment variables
-           . /tmp/configmap.env.values
-           clusterctl upgrade apply --management-group capi-system/cluster-api --contract v1alpha3
-           sleep 30
-           # restore secrets to fill missing secret fields
-           kubectl replace -f /tmp/secrets.with.values.yaml
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    ansible.builtin.command: clusterctl upgrade apply --management-group capi-system/cluster-api --contract v1alpha3
+    environment: "{{ ironic_env | combine(kubeconfig_env) }}"
+    vars:
+      kubeconfig_env:
+        KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+
+  # TODO: Can we check this somehow instead of just waiting?
+  # Relevant issue: https://github.com/kubernetes-sigs/cluster-api/issues/4474
```

Comment:
Maybe we could add a loop to wait for the pods in the capi-system to be available ?

Reply:
Do you know if that will be enough? I saw that there is also some things in `capi-webhook-system`, `capi-kubeadm-control-plane-system` and `capi-kubeadm-bootstrap-system`. Since I'm not really sure what we would have to wait for I just left it as a pause for now, but I can certainly add a wait for some/all of these deployments/pods if you know which are important! :slightly_smiling_face:

Reply:
Ah true, there is also webhook namespace. Maybe we could take the namespaces that starts with `capi*`. But we could it in a separate PR, this should not be a blocker.